runtime/vm/JFRChunkWriter.cpp

```diff
@@ -0,0 +1,641 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2024
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+#include "vm_internal.h"
+
+#include "JFRChunkWriter.hpp"
+
+void
+executionSample(J9VMThread *currentThread)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
+	acquireExclusiveVMAccess(currentThread);
+	do  {
+		if (VM_VMHelpers::threadCanRunJavaCode(walkThread)) {
+			jfrExecutionSample(walkThread);
```

Comment:
This is not strictly legal - the function is expecting the current J9VMThread (though as things stand this will probably work because you have exclusive). Should be refactored to take current and target thread.